pyfarm/jobtypes/core/internals.py

```diff
@@ -83,88 +87,222 @@ class InsufficientSpaceError(Exception):
     pass
 
 
-class Cache(object):
-    """Internal methods for caching job types"""
-    cache = {}
-    JOBTYPE_VERSION_URL = \
-        "%(master_api)s/jobtypes/%(name)s/versions/%(version)s"
-    CACHE_DIRECTORY = config.get("jobtype_cache_directory", "")
+class JobTypeDownloadError(Exception):
+    """
+    Raised when there is some problem download a job type.
+    """
 
-    if not CACHE_DIRECTORY:  # pragma: no cover
-        CACHE_DIRECTORY = None  # make sure it's None
-        logger.warning("Job type cache directory has been disabled.")
 
-    else:
-        try:
-            os.makedirs(CACHE_DIRECTORY)
+class JobTypeNotFound(JobTypeDownloadError):
+    """
+    Raised when we fail to download a job type because
+    it no longer exists on the master.
+    """
+    def __init__(self, name, version):
+        super(JobTypeNotFound, self).__init__(
+            "Job type %s v%s not found" % (name, version))
+        self.name = name
+        self.version = version
 
-        except OSError as e:  # pragma: no cover
-            if e.errno != EEXIST:
-                logger.error(
-                    "Failed to create %r.  Job type caching is "
-                    "now disabled.", CACHE_DIRECTORY)
-                raise
+
+class JobTypeLoader(object):
+    """Class for retrieval, loading and caching of job types."""
+    def __init__(self):
+        if not config["jobtype_enable_cache"]:
+            cache_directory = None
         else:
-            logger.info("Created job type cache directory %r", CACHE_DIRECTORY)
+            try:
+                cache_directory = config["jobtype_cache_directory"]
```

Comment:
Could we join ```config['farm_name']``` to this, if it is set?

That way, if agents get switched between farms for some reason, they won't get the different jobtypes caches mixed up.